README.md
# JSON2Graph Demo

![demo image](./public/images/widget-demo.png?raw=true)

-   **React** 17
-   **tailwindcss:** v3.x
-   **typescript**
-   **SPA, client side rendering**
-   **SigmaJS webGL graph rendering library**
-   **Graphology graph library**
-   **mobx-state-tree state management**

## Get Started

```bash
yarn
# or `npm install`
```

---

## Develop

```bash
yarn dev
# or `npm run dev`
```

---

## Build

```bash
yarn build
# or `npm run build`
```

---

# Overview

Overall the project can be used as a tool to take in relatively unknown dataset rows as JSON data, and create a graph visualization with them.

With graphs the relationship between the data has to be set somewhere, its relatively impossible to look at a dataset and assign node/edge values, because it really depends on what you are looking to visualize. This was a problem that I ran into during my development and research of this project, therefore I thought it would be ideal to create a tool to easily visualize the relationships you were creating.

This web app allows you to define a configuration for a dataset and display the dataset in a graph with that configuration.

Included are a few example datasets, which have been stripped down to under 100k lines of rows as JSON objects. The datasets are reddit comments and yelp businesses, each of which have 2 configurations. Each configuration is showing the data in a different way by setting node/edge values differently.

# Configurations

The config file is located at
`/configs/data.mapping.json`

Configuration definition -

```{
            "id": "yelp_businesses", // required
            "url": "yelp.businesses.json", // required - where to fetch the URL from, along with .env
            "label": "yelp businesses - businesses/states", // optional - what the name of the dataset is in the web app
            "description": "This is an optional description", // optional - a description of the configuration
            // node array contains node object definitions for creating the nodes, you can have as many you would like to visualize
            "nodes": [
                {
                    "name": "business", // required - readable name of the node, doesn't need to be from row data
                    "idAttr": "name", // required - the attribute of the row to be used (what it will use as the ID)
                    "labelAttr": "name", // required - label attribute of the row to be used (what it will look like)
                    "clusterLabel": "Business" // optional - readable cluster or sub label of node, doesn't need to be from row data
                },
                {
                    "name": "State",
                    "idAttr": "state",
                    "labelAttr": "state",
                    "clusterLabel": "State"
                }
            ],
            // edge array contains which nodes you would like to create edges from.
            // These source/target id values need to match the ones set above.
            // The edge array will draw edges from source to target.
            "edges": [
                {
                    "sourceNodeId": "name", // required - the source attribute of the row value used above (name in this case)
                    "targetNodeId": "state", // required - the target attribute of the row value used above(state in this case)
                    "edgeLabel": "Is in", // optional - what to call the edge
                    "merge": true // optional, this will merge in edges that already exist.
                    // Useful for showing certain visualizations where edges don't need to be unique. (yelp dataset)
                }
            ]
        }
```

# Using the App

## Dev Panel

The wrench icon opens a dev panel, which shows you what you are looking at. This has some graph terminology and other stats such as nodes/edges, graph type, edge type and other metrics. It also have some toggles and tools to shape/view the graph.

## Cropping Non Connected Edges

This can be selected and acted on the currently showing dataset. This removes nodes/edges that aren't connected to the larger set.

## Target # of Rows

This will randomly select approximately X number of rows from very large datasets. The included datasets are all 100k rows, and depending on your nodes/edges, can be quite a few data points.

## Simulation

The simulation is configured to run for 5 secs, but you can press the 'play' button to continue the simulation. Ideally in future versions if the dataset was known or 'seen before' we could cache the x/y positions and not have to simulate.

# Responsive Design

This project what built to possibly be used in a dashboard, panel like environment, therefore all controls/visual components are designed with mobile-like space considerations but also useable on desktop.

# Technology Considerations

There was some extensive research into the landscape of current graph visualization libraries. If the shape and size of the dataset isn't generally known, then we should reach for the latest and greatest in displaying the largest sets of data in a browser, which is WebGL. SigmaJS was chosen for this purpose as it utilizes webGL to display graph data backed by graphology JS. The data is ingested, and graph is created with Graphology and then displayed with Sigma. This has proven to be able to show really as many nodes and edges as you would like, but the simulation is very slow. Sigma and this project also utilize web workers to offload simulation from the UI thread. Ideally 10-20k nodes/edges is a good target for speed considerations, but you can simulate the graph as long as you want.

# Caveats

Because WebGL / SigmaJS, showing many iframe instances at a time is a browser constraint and webGL context will be lost. In future versions this can be remedied by reloading the webGL handles programmatically if lost, ideally from the SigmaJS framework.


CHANGELOG.md
#### 0.0.11 (2023-09-22)

##### Build System / Dependencies

* **env:**  prod url update (fa4eaca8)

##### New Features

* **ui:**
  *  Search results box implementation (73e69103)
  *  added changelog modal (c670fdc5)
* **search:**  search implemented stateside (84a80f62)

#### 0.0.9 (2023-09-22)

##### Build System / Dependencies

* **env:**  prod url update (fa4eaca8)

##### New Features

* **ui:**
  *  Search results box implementation (73e69103)
  *  added changelog modal (c670fdc5)
* **search:**  search implemented stateside (84a80f62)

#### 0.0.7 (2023-09-22)

##### Build System / Dependencies

* **env:**  prod url update (fa4eaca8)

##### New Features

* **ui:**
  *  Search results box implementation (73e69103)
  *  added changelog modal (c670fdc5)
* **search:**  search implemented stateside (84a80f62)

#### 0.0.6 (2023-09-21)

##### New Features

* **graph:**  added ability to filter by node degree (e6dea693)

##### Bug Fixes

* **build:**
  *  daisyUI added back to dev deps (e8a7755a)
  *  updates for package and smaller sample sizes (cd40c84f)
* **data:**  updated sampled rows and created smaller sample sets (15f616c6)

##### Other Changes

*  moved graph ref to own store for devtools (be34dd52)

#### 0.0.5 (2023-08-30)

##### Build System / Dependencies

* **cd/cd:**  update for GH build command (666e166c)

##### Chores

* **lint:**
  *  ran rome linter over files, sorted imports (a4e7b936)
  *  ran rome linter over files (de7c81ac)

##### New Features

* **ui:**
  *  added config for hover node functionality (96d27c8d)
  *  update for allowing light and dark mode (906847e1)
* **clustering:**  impl of clustering via max degrees, made configurable (338d22bc)
* **tooltip:**  impl of the tooltip feature (c7492015)
* **state:**
  *  updated to zustood, working through clustering (df1cb2aa)
  *  further refined app to populate from config file on load (04416e06)
  *  updated graph creation code to read from config file and populate graph (c80ab5cd)
* **data:**  added local storage cache (6c8e9032)

##### Bug Fixes

* **build:**
  *  updates for fixing data fetch on prd (82bbbda4)
  *  updates for fixing deployment build (49e453ce)
* **data:**  updated access control headers for data request (bdc804fa)
* **ui:**  small fixes (68bc748a)

##### Other Changes

*  @Azure opensource@microsoft.com (9d969102)
* **ui:**  added testing library, removed sigma references, added layout switching (7c130cf4)

##### Refactors

* **state:**
  *  removed mobx ref, using zustand (0b61c2a8)
  *  updated app/data to use Zustand (6f94f296)
  *  removed eslint, prettier, using rome (d4fe6ebf)

#### 0.0.4 (2022-10-26)

##### Bug Fixes

* **ui:**  fixed overflow on dev panel (27eb7fd0)

#### 0.0.3 (2022-10-26)

##### New Features

* **state:**  updated state management to fetch data via async function, added reload functionality, added the ability to change num rows read (d5def18b)
* **ui:**  tweaked css for small view (e1cb981f)

##### Bug Fixes

* **ui:**
  *  added pulse to status text when simming (0a5d8000)
  *  updated naming scheme for graphRow (90823aa0)
  *  fixed clipping issue on devpanel (4e5e7fc4)

#### 0.0.2 (2022-10-26)

##### Documentation Changes

* **license:**  added MIT license to package.json (d10d82f9)

##### New Features

* **build:**  added changelog roll script and tagging (f014a4f7)
* **ui:**
  *  added tailwind forms (65452bcd)
  *  polish for demo (bae8e565)
  *  added removal toggle for un connected nodes (aa09805b)
  *  closed dev panel on canvas click (80dd88ce)
  *  styled dev panel (68064307)
  *  moved dev panel and added animation (43165775)
  *  styled loading icon (114d1c2f)
  *  removed switch for dev panel button to save space (2237143e)
  *  added package json version to screen for demo (0365da64)
  *  added handling and snack notif for losing webGL context (dc874994)
  *  added centrifuge loading spinner, added error notif tray, added some polish on basic loading screens and animation (1a4bc026)
  *  added a lot of graph/data values to the dev panel and state management (05c2e05c)
  *  added dev panel, customized rendering, adding settings into state management (8574a396)
  *  added state management and dev mode panel as well as styled zoom controls (c6b42db3)
* **data:**  added sample dataset to repo, fixed button onClick (97a50dda)
* **state:**  seperated app and data store (ea291011)
* **demo:**  impl of sigmajs demo code (3c4268e8)

##### Bug Fixes

* **ui:**
  *  fix for prod build ui loading spinner (91a3835e)
  *  updated logo to be slightly off centered (3720de00)
  *  updated zoom min,max linted code and updated click functionality (385168ef)
* **build:**  fixed env values for local/prod (a4b37ac8)

##### Other Changes

* //github.com/Surge-Workshop/surge-web (3f632119)
* //github.com/Surge-Workshop/surge-web (4719e8d9)
*  @Azure opensource@microsoft.com (12222d7e)

##### Refactors

* **settings:**  updated default settings on sigma init for min/max zoom (021557a6)



src/lib/Utils.ts
import IconLoader from "@antv/graphin-icons"
import Graphin from "@antv/graphin"
// Icons from Graphin
const icons = Graphin.registerFontFamily(IconLoader)

import { ComboConfig, EdgeConfig, GraphData } from "@antv/g6"
import { IUserEdge, IUserNode } from "@antv/graphin"
import { store } from "../stores/Store"
import {
    CONTENT_COLORS,
    DataToGraphConfig,
    ExtendedNode,
    NODE_COLORS,
    ThemeColors,
    _EdgeConfig,
    _GraphData,
    _NodeConfig,
} from "./AppTypes"

export const setNodeIcon = (node: IUserNode, nodeConfig: _NodeConfig) => {
    if (!nodeConfig.icon || !node.style) {
        return
    }
    node.style.icon = {
        type: "font",
        fontFamily: "graphin",
        value: icons[nodeConfig.icon as keyof typeof icons],
        fill: getColorFromNodeConfig(nodeConfig, true),
    }
}

export const getColorFromNodeConfig = (
    nodeConfig: _NodeConfig,
    content = false,
) => {
    // find index of the nodeConfig in the list of nodeConfigs
    const index = store.data.dataSet().nodes?.indexOf(nodeConfig)

    // get the color palette
    const themeColors = store.app.colors()

    const source = content ? CONTENT_COLORS : NODE_COLORS

    // get the color from the palette for the nodeConfig index
    const value =
        index !== undefined ? Object.values(source)[index as number] : undefined

    // return the color
    return value !== undefined
        ? themeColors[value as keyof ThemeColors]
        : undefined
}

export const setNodeColor = (node: IUserNode, nodeConfig: _NodeConfig) => {
    if (!nodeConfig || !node.style) {
        return
    }

    node.style.keyshape = {
        ...node.style.keyshape,
        fill: nodeConfig.color
            ? nodeConfig.color
            : getColorFromNodeConfig(nodeConfig),
        fillOpacity: 0.8,
    }
}

function setSizeBasedOnDegrees(graphData: GraphData) {
    const maxDegree = Math.max(
        ...(graphData.nodes?.map((node) =>
            calculateDegree(graphData, node.id),
        ) || []),
    )

    const minSize = 10
    const maxSize = maxDegree * 10

    graphData.nodes?.forEach((node) => {
        // console.log(node.clusterLabel.length)
        const size =
            minSize +
            (maxSize - minSize) *
                (calculateDegree(graphData, node.id) / maxDegree)
        node.style = {
            ...node.style,
            keepVisualSize: true,
            keyshape: {
                ...node.style?.keyshape,
                size,
            },
            label: {
                // @ts-ignore
                value: truncateString(node?.label?.value ?? "", 10),
            },
            // badges: [
            //     {
            //         position: "RT",
            //         type: "text",
            //         value: node.clusterLabel,
            //         size: [node.clusterLabel.length * 10, 15],
            //         fill: COLORS[node.clusterLabel as string],
            //         color: "#fff",
            //     },
            // ],
        }
        node.size = size
        // node.comboId = node.clusterLabel
    })
}

/**
 * @description Truncate a string to a given length
 * @author Logan Hendershot
 * @date 07/25/2023
 * @param {string} str
 * @param {number} maxLength
 * @returns {*}  {string}
 */
export const truncateString = (str: string, maxLength: number): string => {
    if (!str) return ""

    // if not a string, return the original value
    if (typeof str !== "string") {
        return str
    }

    if (str.length <= maxLength) {
        return str
    }

    const truncatedString = str?.substring(0, maxLength)
    return `${truncatedString}...`
}

/**
 * @description Calculate the degree of a node in a graph given the graph data and the node id
 * @author Logan Hendershot
 * @date 07/25/2023
 * @param {{ nodes: IUserNode[]; edges: IUserEdge[] }} graphData
 * @param {string} nodeId
 * @returns {*}  {number}
 */
const calculateDegree = (graphData: GraphData, nodeId: string): number => {
    let degree = 0
    graphData.edges?.forEach((edge) => {
        if (edge.source === nodeId || edge.target === nodeId) {
            degree++
        }
    })
    return degree
}

export const populateGraphinData = (
    data: unknown[],
    config: DataToGraphConfig,
): _GraphData => {
    const graphData = {
        nodes: [] as ExtendedNode[],
        edges: [] as IUserEdge[],
        combos: [] as ComboConfig[],
    }

    console.log("configs", config)

    // populate nodes and edges
    data.forEach((row: unknown) => {
        console.log("row", row)
        config.nodes?.forEach((nodeConfig) => {
            console.log("nodeConfig", nodeConfig)
            addNodeToG6Graph(graphData, row, nodeConfig)
        })

        config.edges?.forEach((edgeConfig) => {
            console.log("edgeConfig", edgeConfig)
            addEdgesToG6Graph(graphData, row, edgeConfig)
        })
    })

    setSizeBasedOnDegrees(graphData)

    // add combos if enabled
    if (store.graph.clusteringEnabled()) {
        setCombos(graphData)
    }

    // return graph
    return graphData
}

/**
 * @description Sets the combos for the graph based on the clustering limit and the number of nodes in each cluster
 * @author Logan Hendershot
 * @date 08/22/2023
 * @param {GraphData} graphData
 */
export const setCombos = (graphData: GraphData) => {
    const combos: ComboConfig[] = []
    // create a map to count the number of nodes in each cluster
    const comboCountMap: { [comboId: string]: number } = {}

    // get the clustering limit from state
    const clusterLimit = store.graph.clusteringLimit()

    // count and sort the number of nodes in each cluster
    graphData.nodes?.forEach((node) => {
        const comboId = node.comboId || ""
        if (!comboCountMap[comboId]) {
            comboCountMap[comboId] = 0
        }
        comboCountMap[comboId]++
    })

    const sortedComboIds = Object.keys(comboCountMap).sort(
        (a, b) => comboCountMap[b] - comboCountMap[a],
    )

    // create a combo for each cluster, up to the clustering limit
    // with a label and a color
    for (let i = 0; i < Math.min(sortedComboIds.length, clusterLimit); i++) {
        const comboId = sortedComboIds[i]
        combos.push({
            id: comboId,
            label: comboId,
            collapsed: false,
            labelCfg: {
                position: "top",
                refY: 10,
                style: {
                    fontSize: 35,
                },
            },
            style: {
                opacity: 0.4,
                fill: getRandomColor(),
            },
        })
    }

    graphData.combos = combos
}

const getRandomColor = (): string => {
    const BLUE_PALLETS = [
        "#e8f6ff",
        "#d5efff",
        "#b3dfff",
        "#86c6ff",
        "#56a0ff",
        "#3078ff",
        "#0d4dff",
        "#0342ff",
        "#0636bc",
        "#10399f",
        "#0a205c",
    ]
    const combinedColors = [...BLUE_PALLETS]
    const randomIndex = Math.floor(Math.random() * combinedColors.length)
    return combinedColors[randomIndex]
}

// create a type that extends IUserNode and includes other properties
// the other properties will be a key value pair of the data from the original data set

export const addNodeToG6Graph = (
    graphData: _GraphData,
    row: any,
    nodeConfig: _NodeConfig,
): void => {
    const record = row as Record<string, string>

    if (!record[nodeConfig.id_data_property]) {
        console.log("unable to add node", nodeConfig.id_data_property)
        return
        // throw new Error("Unable to find property with id attribute given.")
    }

    const nodeExists = graphData.nodes.some(
        (node) => node.id === record[nodeConfig.id_data_property as string],
    )

    // console.log("nodeExists", nodeExists)

    if (!nodeExists) {
        const node: ExtendedNode = {
            id: record[nodeConfig.id_data_property as string],
            ...(nodeConfig.content_data_property && {
                label: {
                    value: row[nodeConfig.content_data_property],
                },
            }),
            style: {
                ...(nodeConfig.content_data_property && {
                    label: {
                        value: row[nodeConfig.content_data_property],
                    },
                }),
                ...(nodeConfig.label && {
                    lavel: nodeConfig.clusterLabel,
                    // badges: [
                    //     {
                    //         position: "RT",
                    //         type: "text",
                    //         value: nodeConfig.clusterLabel,
                    //         size: [15, 15],
                    //         fill: "red",
                    //         color: "#fff",
                    //     },
                    // ],
                }),
            },

            // comboId for cluster combo layout
            comboId: row.subreddit,

            // _metadata values used throughout the app
            _metadata: {
                _data: row,
                // _type
                ...(nodeConfig.label && {
                    _type: nodeConfig.label,
                }),

                // _title
                ...(nodeConfig.content_data_property && {
                    _title: row[nodeConfig.content_data_property],
                }),

                // _title
                ...(nodeConfig.content_data_property && {
                    _body: row[nodeConfig.content_data_property],
                }),

                // // _subtitle
                // ...(nodeConfig.tagAttr && {
                //     _subtitle: row[nodeConfig.tagAttr],
                // }),
                // _clusterLabel
                // _clusterId
                _clusterId: row.subreddit,
            },
        }
        setNodeIcon(node, nodeConfig)
        setNodeColor(node, nodeConfig)
        // console.log("node", node);
        graphData.nodes.push(node)
    }
}

export const addEdgesToG6Graph = (
    graphData: _GraphData,
    row: unknown,
    edgeConfig: _EdgeConfig,
): void => {
    const record = row as Record<string, unknown>

    if (
        !record[edgeConfig.source_node_id] ||
        !record[edgeConfig.target_node_id]
    ) {
        console.log(
            "unable to add edge",
            edgeConfig.source_node_id,
            edgeConfig.target_node_id,
        )
        return
        // throw new Error("Required edge params missing.")
    }

    const edgeExists = graphData.edges?.some(
        (edge) =>
            edge.source === (record[edgeConfig.source_node_id] as string) &&
            edge.target === (record[edgeConfig.target_node_id] as string),
    )

    if (!edgeExists) {
        graphData.edges?.push({
            id: `${record[edgeConfig.source_node_id]}-${
                record[edgeConfig.target_node_id]
            }`,
            source: record[edgeConfig.source_node_id] as string,
            target: record[edgeConfig.target_node_id] as string,
            label: edgeConfig.edgeLabel
                ? record[edgeConfig.edgeLabel as string]
                : undefined,
        })
    }
}

export const generateUniqueId = () => {
    const timestamp = Date.now().toString(36)
    const random = Math.random().toString(36).substr(2, 5)
    return `${timestamp}-${random}`
}

// method to strip out everything before the equals sign and the ]
export const stripAndCamelCase = (obj: { [index: string]: any }) => {
    return Object.entries(obj).reduce((result: any, [key, value]) => {
        let cleanedKey = key.replace(/^\[data-theme=(.+)\]$/, "$1") // Stripping out the non-useful bits of the keys
        // Manually transforming kebab-case to camelCase
        cleanedKey = cleanedKey.includes("-")
            ? cleanedKey
                  .toLowerCase()
                  .split("-")
                  .map((word, idx) =>
                      idx !== 0
                          ? word.charAt(0).toUpperCase() + word.slice(1)
                          : word,
                  )
                  .join("")
            : cleanedKey
        if (value && typeof value === "object") {
            result[cleanedKey] = stripAndCamelCase(value) // Recursion to handle nested objects
        } else {
            result[cleanedKey] = value
        }
        return result
    }, {})
}
